google_guest_agent/snapshot_listener.go

```diff
@@ -0,0 +1,239 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+	"syscall"
+	"time"
+
+	pb "github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/protos"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+	"github.com/hashicorp/golang-lru"
+	"google.golang.org/grpc"
+)
+
+var (
+	seenPreSnapshotOperationIds, _  = lru.New(128)
+	seenPostSnapshotOperationIds, _ = lru.New(128)
+)
+
+type snapshotConfig struct {
+	Timeout               int // seconds
+	ContinueOnScriptError bool
+	PreSnapshotScriptURL  string
+	PostSnapshotScriptURL string
+	Enabled               bool
+}
+
+type invalidSnapshotConfig struct {
+	msg string
+}
+
+func (e *invalidSnapshotConfig) Error() string {
+	return fmt.Sprintf("invalid config: %s", e.msg)
+}
+
+func getSnapshotConfig() (snapshotConfig, error) {
+	cfgfile := configPath
+	if runtime.GOOS == "windows" {
+		cfgfile = winConfigPath
+	}
+
+	config, err := parseConfig(cfgfile)
+	if err != nil && !os.IsNotExist(err) {
+		logger.Errorf("Error parsing config %s: %s", cfgfile, err)
+	}
```

Comment:
these are general configs and not snapshot-specific, should be handled in main as before and these functions should assume this part is done